### PR TITLE
Add failing test for issue #7062 - BasicEntityPersister confuses association types with PDO datatypes

### PR DESCRIPTION
**Pull request by mariusklocke**
Providing a failing testcase as requested in issue #7062

### CODE REVIEW

**Comment by Ocramius** [comment]
@mariusklocke maybe you intended to send this to `2.6`?

**Comment by mariusklocke** [comment]
Changed to 2.6
Got a little confused when reading the contribution guidelines :)

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC7062Test.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        $ranking = $this->createInitialRankingWithRelatedEntities();
+        $this->modifyRanking($ranking);
+
+        // Clear the EntityManager to make sure we fetch entities from database
+        $this->_em->clear();
+
+        // Now try to read entities again and check if all changes have been made persistent correctly
+        /** @var Season $season */
+        $season = $this->_em->find(Season::class, self::SEASON_ID);
+        $this->assertInstanceOf(Season::class, $season);
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC7062Test.php]: Please use `self::assert*` instead since the method is static.

**Review by lcobucci — CHANGES_REQUESTED** [review]
@mariusklocke can you please combine the models in the test file just like https://github.com/doctrine/doctrine2/blob/30a063ef9d7b14d0c74a192d87dc31c600503a4b/tests/Doctrine/Tests/ORM/Functional/Ticket/GH6531Test.php ? 

This would also make things easier for us to review, you to fix coding standard issues using `phpcbf`, and also removes all the changes you've made in `OrmFunctionalTestCase` (which are not really necessary).

**Comment by mariusklocke** [comment]
Thanks for your quick feedback. Will do that in the next days.

**Comment by lcobucci** [comment]
@mariusklocke @Ocramius wow, this is a tough one! The cause of the issue is that for some reason the ORM is not using the correct type for the `season_id` on the update query... I'll continue my investigation here

**Comment by mariusklocke** [comment]
@lcobucci Did you read my analysis where i reported the [issue](https://github.com/doctrine/doctrine2/issues/7062)? Can you state if my assumptions there are correct?

**Review by lcobucci — APPROVED** [review]
@mariusklocke I've changed things a bit in the test to always fetch things from the database, just to ensure that things are isolated.

The fix itself was pretty simple (with a little help from @guilhermeblanco).

Thanks for your contribution!

**Comment by lcobucci** [comment]
> @lcobucci Did you read my analysis where i reported the issue? Can you state if my assumptions there are correct?

@mariusklocke I read it, after I've posted my comment... you're assumptions are indeed correct

**Comment by mariusklocke** [comment]
I'm glad you found a fix. Debugging that took me a while :)

**Comment by lcobucci** [comment]
@mariusklocke the only thing is that your mapping quite complex and it uncovers some scenarios that we haven't thought about (the fact that `RankingPosition` has `Ranking` as id, but that id is actually `Season` is making the `IdentifierFlattener` a bit crazy - only when running with L2C enabled)...

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH7062Test.php [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+ *
+ * @Entity
+ * @Table(name="soccer_rankings")
+ */
+class GH7062Ranking
+{
+    /**
+     * @Id
+     * @OneToOne(targetEntity=GH7062Season::class, inversedBy="ranking")
+     * @JoinColumn(name="season", referencedColumnName="id")
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH7062Test.php]: @Ocramius @guilhermeblanco this test revealed a nice unrelated issue (when loading things with L2C the identifier flattener tries to use the [column name in an array that doesn't have that index](https://travis-ci.org/doctrine/doctrine2/jobs/346051424#L959)).
Forcing the entity to use the name of the field as the name of the column was the only way to isolate the issue that @mariusklocke described... what are your thoughts about this workaround?

Comment (guilhermeblanco) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/GH7062Test.php]: @lcobucci as long as we can open a new issue with the new problem, I'm good to isolate issues in 2 separate PRs. =)

Comment (lcobucci) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/GH7062Test.php]: @guilhermeblanco that's the idea

Comment (Ocramius) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/GH7062Test.php]: Yeah, that's a separate issue - let's not creep it into here.

**Review by Ocramius — CHANGES_REQUESTED** [review]
I'm fine with the patch overall, but the test uses a DSL from a different domain, and that makes things quite complex to follow when reasoning about ORM details.

I'd rather get rid of all the public methods in these entities, and modify all states via `public` properties instead, making the classes ORM-naming-specific instead of flavored with `Season` and `Team`

**Comment by lcobucci** [comment]
@Ocramius how about now?

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH7062Test.php (hunk 2 above)
Comment (lcobucci) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/GH7062Test.php]: https://github.com/doctrine/doctrine2/issues/7096 created to handle that

**Comment by Ocramius** [comment]
@lcobucci if the test still fails with the patch not applied, then :+1:

I'd rather completely get rid of the team/season/etc thing...

**Comment by lcobucci** [comment]
> @lcobucci if the test still fails with the patch not applied, then 👍

@Ocramius it does fail.

> I'd rather completely get rid of the team/season/etc thing...

I removed all the methods but decided to keep the whole thing due to the complicated setup of FKs (which is exactly what affects the bug).

**Review by Ocramius — APPROVED** [review]

**Comment by Ocramius** [comment]
:ship:

**Comment by Ocramius** [comment]
Btw, just noticed that recent patches aren't in `2.7`. Will need to merge `2.6` up (not nice, but not as important as `master` right now)

**Comment by Ocramius** [comment]
Also cherry-picking this into `master`

**Comment by Ocramius** [comment]
Interestingly, the test passes on `master` without actually needing to apply the patch - accidentally fixed by fixing our mapping system :-)

**Comment by Ocramius** [comment]
In `master` as per doctrine/doctrine2@bbc1ed93b14e079cd403cb5b2b1134da35af431e
In `2.7` as per doctrine/doctrine2@4f6d47b88a84896dadaadd18212df6bb504f2e97

**Pull request merged**